writer.go

```diff
@@ -0,0 +1,565 @@
+	// Length is the number of items in the table.
+	Length uint64
+}
+
+func (t *tableDef) marshal(pathDir string) error {
+	t.DBVersion = dbVersion
+	b, err := json.Marshal(t)
+	if err != nil {
+		return err
+	}
+	if err := os.WriteFile(filepath.Join(pathDir, "tabledef.json"), b, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *tableDef) unmarshal(pathDir string) error {
+	r, err := os.Open(filepath.Join(pathDir, "tabledef.json"))
+	if err != nil {
+		return err
+	}
+
+	dec := json.NewDecoder(r)
+	dec.DisallowUnknownFields()
+
+	if err := dec.Decode(t); err != nil {
+		unsupported := &json.UnsupportedValueError{}
+		if errors.Is(err, unsupported) {
+			return fmt.Errorf("current tableDef does not support a found field: %w", err)
+		}
+		return err
+	}
+	return nil
+}
+
+type columnDef struct {
+	// Name is the name of index column.
+	Name string
+	// Path is the path to that BadgerDB.
+	Path string
+	// AllowDuplicates tells if the index allows duplicate keys.
+	AllowDuplicates bool
+}
+
+// Writer represents our disk database.
+type Writer struct {
+	// dirPath is the path to the root directory for this database.
+	dirPath string
+	// inmemory indicates to run the DB in memory instead of persistence to disk.
+	inmemory bool
+	// primary is the primary data table.
+	primary *Index
+	// indexes are the set of indexes by name.
+	indexes map[string]*Index
+	// indexList is the indexes in the order they were added.
+	indexList []*Index
+
+	// wg is the number or writes that are happening.
+	wg sync.WaitGroup
+
+	// writeCh gives us access to a writer goroutine pool.
+	writeCh chan func()
+
+	logger badger.Logger
+}
+
+// Index represents an index on our databse.
+type Index struct {
+	// Name of the index. This must be unique.
+	Name string
+	// AllowDuplicates indicates if this index allows duplicate keys for the index.
+	AllowDuplicates bool
+
+	// path is the path to the index.
+	path string
+
+	// db is access to the database for this index.
+	db *badger.DB
+
+	toWrite [][2][]byte
+
+	// bufferPool has a set of buffers for resuse when writing to this index.
+	buffPool chan *z.Buffer
+
+	// counter is a number used in the badgerDB key when this is the primary key or
+	// Name == "primary".
+	counter atomic.Uint64
+
+	// mu protects keyCount.
+	mu sync.Mutex
+
+	// keyCount keeps count of how many times a particular value has been seen
+	// so that we can increment a counter when writing the key to badgerdb. Aka
+	// so you can get "samekey.1", "samekey.2", ... map key is string because you
+	// can't use []byte. Only used if AllowDuplicates is true.
+	keyCount map[string]*atomic.Uint64
+}
+
+// type WriteOption func(db *Writer) error
+
+// WriteOption is optional arguments for New().
+type WriteOption interface {
+	writeOption()
+}
+
+// Indexes provide the indexes that will be used on this database. Can be used with:
+//   - New()
+func WithIndexes(indexes ...*Index) interface {
+	WriteOption
+	calloptions.CallOption
+} {
```

Comment:
why not define this interface explicitly? it's also used below in `WithInMemory`, and could be used as a subset of the returned `interface` of `WithLogger`

Reply:
My first reason is that it obfuscate's the code.  If you look a this and understand the pattern, you can immediately tell a few things:

* it can be used with WriteOption
* it is a CallOption

If I abstract that, it becomes less readable as the further matching options go farther down in the file away from the interface.

The second becomes the naming of the interface.  Right now this would be something like WriteCallOption and the WithLogger() would have another called WriteOpenCallOption.

Now say I create a new method that now wants to use WithIndexes(), I'd have to either create a new interface that imports WriteCallOption and has another name like WriteBulkCallOption, which gets ugly.  

To me that starts to look a lot uglier.

Reply:
if you think it obfuscates, then fair enough